File: encoders/saeed/crypto/md4.cpp
#include "md4.h"

MD4::MD4(int rnds, int dobbertin, int bits, bool initBlock)
    : MDHash(16, 4, rnds, initBlock)
{
    this->dobbertin = dobbertin == 1;
    this->bits = bits;
}

void MD4::encode()
{
    /* Message words */
    for (int i = 0; i < 16; i++)
        cnf.newVars(w[i], 32, "w" + to_string(i));

    for (int i = 0; i < 4; i++)
        cnf.newVars(chain[i], 32);

    for (int i = 0; i < 4; i++)
        cnf.newVars(out[i], 32, "hash" + to_string(i));

    if (initialBlock) {
        cnf.fixedValue(chain[0], 0x67452301); // A
        cnf.fixedValue(chain[1], 0xefcdab89); // B
        cnf.fixedValue(chain[2], 0x98badcfe); // C
        cnf.fixedValue(chain[3], 0x10325476); // D
    }

    /* Round constants */
    Word k[2];
    cnf.newVars(k[0], 32);
    cnf.newVars(k[1], 32);
    cnf.fixedValue(k[0], 0x5a827999);
    cnf.fixedValue(k[1], 0x6ed9eba1);

    /* Round rotations */
    int s[48] = {
        3,
        7,
        11,
        19,
        3,
        7,
        11,
        19,
        3,
        7,
        11,
        19,
        3,
        7,
        11,
        19,
        3,
        5,
        9,
        13,
        3,
        5,
        9,
        13,
        3,
        5,
        9,
        13,
        3,
        5,
        9,
        13,
        3,
        9,
        11,
        15,
        3,
        9,
        11,
        15,
        3,
        9,
        11,
        15,
        3,
        9,
        11,
        15,
    };

    /* Round message word indices */
    int ind[48] = {
        0,
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        0,
        4,
        8,
        12,
        1,
        5,
        9,
        13,
        2,
        6,
        10,
        14,
        3,
        7,
        11,
        15,
        0,
        8,
        4,
        12,
        2,
        10,
        6,
        14,
        1,
        9,
        5,
        13,
        3,
        11,
        7,
        15,
    };

    cnf.assign(q[0], chain[0]); // A
    cnf.assign(q[3], chain[1]); // B
    cnf.assign(q[2], chain[2]); // C
    cnf.assign(q[1], chain[3]); // D

    /* Main loop */
    for (int i = 0; i < rounds; i++) {
        Word t;
        cnf.newVars(t, 32);

        Word f;
        cnf.newVars(f, 32);

        if (i < 16) {
            cnf.ch(f, q[i + 3], q[i + 2], q[i + 1]);
            cnf.add3(t, q[i], f, w[ind[i]]);
        } else if (i < 32) {
            cnf.maj3(f, q[i + 3], q[i + 2], q[i + 1]);
            cnf.add4(t, q[i], f, w[ind[i]], k[0]);
        } else {
            cnf.xor3(f, q[i + 3], q[i + 2], q[i + 1]);
            cnf.add4(t, q[i], f, w[ind[i]], k[1]);
        }

        cnf.rotl(q[i + 4], t, s[i]);
    }

    // Dobbertin's constraints
    if (dobbertin) {
        // Dobbertin's constant
        unsigned int k = 0xffffffff;

        int q_indices[12] = {
            16, 20, 24, 28, // q[16] = Q[13] (from [Debapratim De et al. 2007])
            17, 21, 25, 29,
            18, 22, 26, 30
        };
        // Index of q that needs relaxation
        int p = 16;
        for (int& i : q_indices) {
            if (i == p && bits != 32) {
                // TODO: Offset the value of k
                cnf.fixedValue(q[i] + (32 - bits), k, bits);
            } else {
                cnf.fixedValue(q[i], k);
            }
        }
    }

    int R = rounds, r = rounds % 4;
    cnf.add2(out[0], chain[0], q[R + (4 - r) % 4]);
    cnf.add2(out[1], chain[1], q[R + (3 - r) % 4]);
    cnf.add2(out[2], chain[2], q[R + (6 - r) % 4]);
    cnf.add2(out[3], chain[3], q[R + (5 - r) % 4]);
}

File: encoders/nejati-preimage/src/md4.rs
use nejati_sat::{Formula, Word};

const ROTATIONS: [usize; 48] = [
    3, 7, 11, 19, 3, 7, 11, 19, 3, 7, 11, 19, 3, 7, 11, 19, 3, 5, 9, 13, 3, 5, 9, 13, 3, 5, 9, 13,
    3, 5, 9, 13, 3, 9, 11, 15, 3, 9, 11, 15, 3, 9, 11, 15, 3, 9, 11, 15,
];

const MESSAGE_INDEX: [usize; 48] = [
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 0, 4, 8, 12, 1, 5, 9, 13, 2, 6, 10, 14,
    3, 7, 11, 15, 0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15,
];

pub struct Md4Encoding {
    pub formula: Formula,
    pub message: [Word; 16],
    output: [Word; 4],
}

impl Md4Encoding {
    pub fn build(
        rounds: usize,
        use_xor_clauses: bool,
        dobbertin: bool,
        relaxed_bits: usize,
    ) -> Self {
        assert!((1..=48).contains(&rounds));
        assert!((1..=32).contains(&relaxed_bits));
        assert!(!dobbertin || rounds >= 27);

        let mut formula = Formula::new(use_xor_clauses);
        let message: [Word; 16] =
            std::array::from_fn(|word| formula.new_word(Some(format!("w{word}"))));
        let chain: [Word; 4] = std::array::from_fn(|_| formula.new_word(None));
        let output: [Word; 4] =
            std::array::from_fn(|word| formula.new_word(Some(format!("hash{word}"))));

        formula.fixed_value(&chain[0], 0x6745_2301);
        formula.fixed_value(&chain[1], 0xefcd_ab89);
        formula.fixed_value(&chain[2], 0x98ba_dcfe);
        formula.fixed_value(&chain[3], 0x1032_5476);

        let round_constant_2 = formula.new_word(None);
        let round_constant_3 = formula.new_word(None);
        formula.fixed_value(&round_constant_2, 0x5a82_7999);
        formula.fixed_value(&round_constant_3, 0x6ed9_eba1);

        let mut q = vec![chain[0], chain[3], chain[2], chain[1]];
        for round in 0..rounds {
            let temporary = formula.new_word(None);
            let function = formula.new_word(None);
            if round < 16 {
                formula.choose(&function, &q[round + 3], &q[round + 2], &q[round + 1]);
                formula.add3(
                    &temporary,
                    &q[round],
                    &function,
                    &message[MESSAGE_INDEX[round]],
                );
            } else if round < 32 {
                formula.majority3(&function, &q[round + 3], &q[round + 2], &q[round + 1]);
                formula.add4(
                    &temporary,
                    &q[round],
                    &function,
                    &message[MESSAGE_INDEX[round]],
                    &round_constant_2,
                );
            } else {
                formula.xor3(&function, &q[round + 3], &q[round + 2], &q[round + 1]);
                formula.add4(
                    &temporary,
                    &q[round],
                    &function,
                    &message[MESSAGE_INDEX[round]],
                    &round_constant_3,
                );
            }
            q.push(formula.rotate_left(&temporary, ROTATIONS[round]));
        }

        if dobbertin {
            for index in [16usize, 20, 24, 28, 17, 21, 25, 29, 18, 22, 26, 30] {
                if index == 16 && relaxed_bits != 32 {
                    formula.fixed_value_bits(
                        &q[index][32 - relaxed_bits..],
                        u32::MAX,
                        relaxed_bits,
                    );
                } else {
                    formula.fixed_value(&q[index], u32::MAX);
                }
            }
        }

        let remainder = rounds % 4;
        formula.equivalent(&output[0], &q[rounds + (4 - remainder) % 4]);
        formula.equivalent(&output[1], &q[rounds + (3 - remainder) % 4]);
        formula.equivalent(&output[2], &q[rounds + (6 - remainder) % 4]);
        formula.equivalent(&output[3], &q[rounds + (5 - remainder) % 4]);

        Self {
            formula,
            message,
            output,
        }
    }

    pub fn fix_output(&mut self, target: [u32; 4]) {
        for (word, value) in self.output.iter().zip(target) {
            self.formula.fixed_value(word, value);
        }
    }

    pub fn fix_message_prefix(&mut self, message: &[u32; 16], bits: usize) {
        for bit in 0..bits {
            let word = bit / 32;
            let offset = bit % 32;
            let value = (message[word] >> offset) & 1;
            let variable = self.message[word][offset];
            self.formula
                .add_clause(vec![if value == 1 { variable } else { -variable }]);
        }
    }
}

pub fn compression_state(message: &[u32; 16], rounds: usize) -> [u32; 4] {
    let mut state = [0x6745_2301u32, 0xefcd_ab89, 0x98ba_dcfe, 0x1032_5476];
    for round in 0..rounds {
        let slot = (48 - round) % 4;
        let b = state[(slot + 1) % 4];
        let c = state[(slot + 2) % 4];
        let d = state[(slot + 3) % 4];
        let (function, constant) = if round < 16 {
            ((b & c) | ((!b) & d), 0)
        } else if round < 32 {
            ((b & c) | (b & d) | (c & d), 0x5a82_7999)
        } else {
            (b ^ c ^ d, 0x6ed9_eba1)
        };
        state[slot] = state[slot]
            .wrapping_add(function)
            .wrapping_add(message[MESSAGE_INDEX[round]])
            .wrapping_add(constant)
            .rotate_left(ROTATIONS[round] as u32);
    }
    state
}
